src/dmd/dcast.d

```diff
@@ -1508,6 +1487,32 @@ Expression castTo(Expression e, Scope* sc, Type t)
             this.t = t;
         }
 
+        // Try casting the alias this member. Return the expression if it succeeds, null otherwise.
+        private Expression tryAliasThisCast(Type tob, Type t1b)
+        {
+            AggregateDeclaration t1ad = isAggregate(t1b);
+            if (!t1ad)
+                return null;
+
+            AggregateDeclaration toad = isAggregate(tob);
+            if (t1ad == toad || !t1ad.aliasthis)
+                return null;
+
+            if (att1 && t1b.equivalent(att1))
+                return null;
+            else if (!att1 && t1b.checkAliasThisRec())
+                att1 = t1b;
```

Comment:
@WalterBright  Lines 1501-1504 were not previously in this function.

Reply:
These are the check needed before any `alias this` resolving to avoid infinite recursion.